pmlc/dialect/pxa/transforms/reorder_layout.cc

```diff
@@ -0,0 +1,264 @@
+/// Output:
+///   reorder map (B)  = (d0, d1) -> (d0, d1 floordiv 8 floordiv 2, d1 floordiv
+///   8 % 2, 0) reordered shape  = <7, 6, 2, 8> reordered vector = <1, 1, 1, 8>
+///
+/// Note: to obtain affine map from input space to expanded space composition
+///       A o B can be used (with simplification).
+/// A o B = (d0, d1) -> (d0 + d1, d0, d1, 0)
+ReorderDesc expandAffineMap(mlir::AffineMap map, mlir::ArrayRef<uint64_t> shape,
+                            mlir::ArrayRef<uint64_t> vector,
+                            mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  auto reorderMap = mlir::AffineMap::getMultiDimIdentityMap(map.getNumResults(),
+                                                            map.getContext());
+  mlir::SmallVector<uint64_t, 6> reorderedShape(shape.begin(), shape.end());
+  mlir::SmallVector<uint64_t, 6> reorderedVector(vector.begin(), vector.end());
+  return ReorderDesc{reorderMap, reorderedShape, reorderedVector};
+}
+
+/// Create affine permutation map that sorts resulting space dimensions in order
+/// of increasing schedule.
+///
+/// Input:
+///   map (A)    = (d0, d1) -> (d0 + d1, d0, d1, 0)
+///   shape      = <7, 6, 2, 8>
+///   vector     = <1, 1, 1, 8>
+///   schedule   = <1, 0>
+/// Output:
+///   reorder map (B)  = (d0, d1, d2, d3) -> (d2, d0, d1, d3)
+///   reordered shape  = <2, 7, 6, 8>
+///   reordered vector = <1, 1, 1, 8>
+///
+/// Note: to obtain affine map from input space to sorted space composition
+///       A o B can be used.
+///   A o B = (d0, d1) -> (d1, d0 + d1, d0)
+ReorderDesc sortAffineMap(mlir::AffineMap map, mlir::ArrayRef<uint64_t> shape,
+                          mlir::ArrayRef<uint64_t> vector,
+                          mlir::ArrayRef<unsigned> schedule) {
+  // TODO: Implement.
+  auto reorderMap = mlir::AffineMap::getMultiDimIdentityMap(map.getNumResults(),
+                                                            map.getContext());
+  mlir::SmallVector<uint64_t, 6> reorderedShape(shape.begin(), shape.end());
+  mlir::SmallVector<uint64_t, 6> reorderedVector(vector.begin(), vector.end());
+  return ReorderDesc{reorderMap, reorderedShape, reorderedVector};
+}
+
+/// Tile affine map using integral constraints to optimize specified schedule.
+/// Returns llvm::None if current affine map already is optimal.
+///
+/// Input:
+///   map (A)           = (d0, d1) -> (d0 + d1, d0 * 16 + d1 * 8)
+///   shape        = <7, 96>
+///   vector       = <1, 8>
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+///   schedule    = <1, 0>
+/// Output:
+///   reorder map = (d0, d1) -> (d1 % 16, d0, d1 floordiv 16)
+///   shape map = (d0, d1) -> (16, d0, d1 ceildiv 16)
+///   vector map = (d0, d1) -> (1, 1)
+///
+/// Note: to obtain affine map from input space to tiled space composition
+///       A o B can be used (with simplification).
+///   A o B = (d0, d1) -> (d1, d0 + d1, d0)
+mlir::Optional<ReorderDesc> tileAffineMap(mlir::AffineMap map,
```

Comment:
The example seems to have errors in it or perhaps I just don't follow (why is vector map's output 2d?). Also based on the building blocks above, it seems like we would split the lower dimension into 3 bits, and then reorder, so we should have a 4d output?

Reply:
Sorry, I forgot to update after some changes, fixed. The result of this operation should be just composition of expand and sort.